node/network/src/proxy/mod.rs

```diff
@@ -0,0 +1,117 @@
+use std::sync::Arc;
+
+use futures::Poll;
+use futures::stream::Stream;
+use futures::sync::mpsc::Receiver;
+use tokio::prelude::Async;
+
+use primitives::serialize::Encode;
+
+use crate::peer::PeerMessage;
+use crate::protocol::{forward_msg, PackedMessage};
+use crate::protocol::Package;
+
+mod registry;
+
+/// Proxy Handlers implementations
+pub mod benchmark;
+pub mod debug;
+pub mod dropout;
+pub mod throttling;
+
+pub struct Proxy {
+    handlers: Vec<Arc<ProxyHandler>>,
+}
+
+/// All messages before being sent to other participants will be processed by several handlers
+/// chained sequentially in a pipeline.
+///
+/// Each handler will receive a stream of messages and will return (not necessarily the same)
+/// stream of messages. This method allows easy manipulation of messages before being emitted
+/// to other authorities through the network.
+///
+/// The order in which handlers are grouped can affect the behavior of the proxy.
+impl Proxy {
+    pub fn new(handlers: Vec<Arc<ProxyHandler>>) -> Self {
+        Self {
+            handlers,
+        }
+    }
+
+    /// Spawn proxies, start task that polls messages and pass them through proxy handlers.
+    pub fn spawn(&mut self, inc_messages: Receiver<PackedMessage>) {
+        let mut stream: Box<Stream<Item=Package, Error=()> + Send + Sync> = Box::new(PackedMessageStream::new(inc_messages));
```

Comment:
`PackedMessageStream` is completely unnecessary because it reimplements `flatten`. We should use `flatten` like I suggested in another comment.

Reply:
removed! didn't notice `flatten` and was looking for something like `concat`.